runtime/sys.c

```diff
@@ -286,10 +288,34 @@ CAMLprim value caml_sys_remove(value name)
   CAMLparam1(name);
   char_os * p;
   int ret;
+
   caml_sys_check_path(name);
   p = caml_stat_strdup_to_os(String_val(name));
   caml_enter_blocking_section();
   ret = unlink_os(p);
+#ifdef _WIN32
+  /* On Windows, trying to unlink a symlink to a directory will return
+   * EACCES, but the symlink can be deleted with rmdir. */
+  if (ret != 0 && errno == EACCES) {
+    if ((ret = GetFileAttributes(p)) != INVALID_FILE_ATTRIBUTES) {
+      if ((ret & FILE_ATTRIBUTE_REPARSE_POINT) &&
+          (ret & FILE_ATTRIBUTE_DIRECTORY)) {
+        ret = rmdir_os(p);
+      } else if (ret & FILE_ATTRIBUTE_DIRECTORY) {
+        /* POSIX uses EPERM rather than EACCES when the path is a
+         * directory. */
+        errno = EPERM;
+        ret = -1;
+      } else {
+        errno = EACCES;
+        ret = -1;
+      }
+    } else {
+      errno = EACCES;
+      ret = -1;
+    }
```

Comment:
In the separate PR, if you put the return value of `GetFileAttributes` into a different `DWORD`, then you'd have no need to update `ret` here. `errno` is only clobbered by `rmdir_os`, so there's never a need to set it to `EACCES`.